File: src/lua/scimark2/SparseCompRow.lua
SparseCompRow = class(function(f)
end)

function SparseCompRow.mult(n, cycles, vy, val, row, col, vx)
    for p=1,cycles do
        for r=1,n do
            local sum = 0
            for i=row[r],row[r+1]-1 do sum = sum + vx[col[i]] * val[i] end
            vy[r] = sum
        end
    end
end

File: src/go/entityfx/scimark2/sparseCompRow.go
package scimark2

func sparse_cr_num_flops(N int, nz int, num_iterations int) float64 {
	actual_nz := (nz / N) * N
	return float64(actual_nz) * 2.0 * float64(num_iterations)
}

func sparse_cr_matmult(y []float64, val []float64, row []int, col []int, x []float64, NUM_ITERATIONS int) {
	M := len(row) - 1
	for reps := 0; reps < NUM_ITERATIONS; reps++ {
		for r := 0; r < M; r++ {
			sum := .0
			rowR := row[r]
			rowRp1 := row[r + 1]
			for i := rowR; i < rowRp1; i++ {
				sum += (x[col[i]] * val[i])
			}
			y[r] = sum
		}
	}
}